File: android/app/src/main/kotlin/ai/gowtam/oak/services/AuthService.kt
package ai.gowtam.oak.services

import ai.gowtam.oak.networking.Endpoint
import ai.gowtam.oak.networking.OakApiClient
import ai.gowtam.oak.networking.OakError
import ai.gowtam.oak.networking.TokenStore
import ai.gowtam.oak.wire.AuthVerifyResponse
import ai.gowtam.oak.wire.MeResponse
import android.util.Log
import kotlinx.serialization.Serializable

/**
 * The sign-in lifecycle seam (component-design.md "Services layer"; mirrors iOS
 * `AuthService`). View models depend on this **interface** (never `LiveAuthService`)
 * so they unit-test against a fake.
 *
 * Every method is `suspend` and surfaces failures as the single typed [OakError].
 * Sign-in is passwordless email-OTP: [requestCode] mails a 6-digit code, [verify]
 * exchanges it for a session — storing the raw Bearer token via [TokenStore] on
 * success. [me] reports the current state on launch, [signOut] returns to guest, and
 * [deleteAccount] performs real backend deletion.
 */
interface AuthService {
    /**
     * Requests a one-time code be emailed to [email]. Maps the route's failures to
     * [OakError]: `invalid_email` (400) → [OakError.Http], `rate_limited` (429) →
     * [OakError.RateLimited], `email_failed` (502) → [OakError.Http]. The success path
     * is non-enumerating (identical for new and returning emails), so there is nothing
     * to return.
     */
    suspend fun requestCode(email: String)

    /**
     * Verifies [code] for [email]. On success the raw session token is stored via
     * [TokenStore] (so it survives relaunch and rides every authed request as
     * `Authorization: Bearer`) and the resolved [Account] is returned. A wrong or
     * expired code throws an [OakError] (`invalid_code` / `invalid_or_expired` /
     * `too_many_attempts`).
     */
    suspend fun verify(email: String, code: String): Account

    /**
     * Reports the current auth state for launch restore: the request carries the
     * stored Bearer token (when present); a valid token resolves to
     * [AuthState.SignedIn], an absent or invalid token to [AuthState.Guest] (the route
     * returns guest as a first-class 200, never an error).
     */
    suspend fun me(): AuthState

    /**
     * Ends the device session: best-effort server revoke, then always clears the
     * stored token so the app returns to guest. Idempotent — calling it without a
     * session is a no-op.
     */
    suspend fun signOut()

    /**
     * Permanently deletes the account and all its server data, then clears the local
     * token. A transport/HTTP failure (other than an already-orphaned token)
     * propagates so the UI does not falsely claim deletion.
     */
    suspend fun deleteAccount()
}

/**
 * The result of a successful [AuthService.verify]. [created] distinguishes a
 * first-time signup (`true`) from a returning login (`false`); the UI may greet a
 * new account differently.
 */
data class Account(val email: String, val created: Boolean)

/** Whether the user is browsing as a guest or signed in with an email account. */
sealed interface AuthState {
    data object Guest : AuthState
    data class SignedIn(val email: String) : AuthState
}

/**
 * Production [AuthService] over [OakApiClient] (the network) and [TokenStore]. It
 * owns the only policy decisions the auth surface needs: when to persist the token
 * (on [verify]) and when to clear it (on [signOut] / [deleteAccount]).
 */
class LiveAuthService(
    private val apiClient: OakApiClient,
    private val tokenStore: TokenStore,
) : AuthService {

    override suspend fun requestCode(email: String) {
        val endpoint = Endpoint(
            method = Endpoint.Method.POST,
            path = "/api/auth/request-code",
            body = Endpoint.jsonBody(RequestCodeBody.serializer(), RequestCodeBody(email)),
            requiresAuth = false,
        )
        apiClient.sendNoContent(endpoint)
    }

    override suspend fun verify(email: String, code: String): Account {
        val endpoint = Endpoint(
            method = Endpoint.Method.POST,
            path = "/api/auth/verify",
            body = Endpoint.jsonBody(VerifyBody.serializer(), VerifyBody(email, code)),
            requiresAuth = false,
        )
        val response = apiClient.send(endpoint, AuthVerifyResponse.serializer())
        // Persist the Bearer token — the one place a token is stored.
        tokenStore.set(response.token)
        return Account(email = response.email, created = response.created)
    }

    override suspend fun me(): AuthState {
        val endpoint = Endpoint(method = Endpoint.Method.GET, path = "/api/auth/me", requiresAuth = true)
        val response = apiClient.send(endpoint, MeResponse.serializer())
        val email = response.email
        return if (response.signedIn && email != null) AuthState.SignedIn(email) else AuthState.Guest
    }

    override suspend fun signOut() {
        // Best-effort server revoke: a transport/HTTP failure must NOT block the
        // local return to guest, so it is logged (never swallowed silently) and the
        // token is cleared regardless. Idempotent.
        val endpoint = Endpoint(method = Endpoint.Method.POST, path = "/api/auth/signout", requiresAuth = true)
        try {
            apiClient.sendNoContent(endpoint)
        } catch (e: OakError) {
            Log.e(TAG, "sign-out endpoint failed (${e::class.simpleName}); clearing local token anyway")
        }
        tokenStore.clear()
    }

    override suspend fun deleteAccount() {
        val endpoint = Endpoint(method = Endpoint.Method.DELETE, path = "/api/auth/account", requiresAuth = true)
        try {
            apiClient.sendNoContent(endpoint)
        } catch (e: OakError.Unauthorized) {
            // The token is already orphaned (the account row is gone) — clearing the
            // local token completes the deletion from the device's point of view.
            Log.i(TAG, "account deletion saw 401 (already orphaned); clearing local token")
        }
        // Reached only on a confirmed 2xx delete or an already-orphaned 401: every
        // other failure (transport / 5xx / rate limit) propagated above with the
        // token intact, so the account is genuinely gone before we clear it.
        tokenStore.clear()
    }

    private companion object {
        const val TAG = "Oak.Auth"
    }
}

/** `POST /api/auth/request-code` body. `email` is identical on the wire. */
@Serializable
private data class RequestCodeBody(val email: String)

/** `POST /api/auth/verify` body. `email`/`code` are identical on the wire. */
@Serializable
private data class VerifyBody(val email: String, val code: String)

File: ios/OakApp/Services/AuthService.swift
import Foundation

/// The sign-in lifecycle seam (component-design.md "Services layer";
/// accounts-and-access.md M-ACCT-US-1/2/3/6). View models depend on this
/// **protocol** (never `LiveAuthService`) so they unit-test against `FakeAuthService`.
///
/// Every method is `async throws` and surfaces failures as the single typed
/// ``OakError``. Sign-in is passwordless email-OTP: ``requestCode(email:)`` mails a
/// 6-digit code, ``verify(email:code:)`` exchanges it for a session — storing the
/// raw Bearer token in the Keychain on success (M-BR-ACCT-5). ``me()`` reports the
/// current state on launch, ``signOut()`` returns to guest, and
/// ``deleteAccount()`` performs real backend deletion (M-BR-ACCT-6).
protocol AuthService: Sendable {
  /// Requests a one-time code be emailed to `email`.
  ///
  /// Maps the route's failures to ``OakError``: `invalid_email` (400) →
  /// `.http`, `rate_limited` (429) → `.rateLimited`, `email_failed` (502) →
  /// `.http`. The success path is non-enumerating (identical for new and
  /// returning emails), so there is nothing to return.
  func requestCode(email: String) async throws

  /// Verifies `code` for `email`. On success the raw session token is written to
  /// the Keychain (so it survives relaunch and rides every authed request as
  /// `Authorization: Bearer`) and the resolved ``Account`` is returned. A wrong
  /// or expired code throws an ``OakError`` (`invalid_code` /
  /// `invalid_or_expired` / `too_many_attempts`).
  func verify(email: String, code: String) async throws -> Account

  /// Reports the current auth state for launch restore: the request carries the
  /// stored Bearer token (when present); a valid token resolves to
  /// `.signedIn(email:)` with an optional `lastUsedScope` (the account's
  /// remembered new-chat default), an absent or invalid token to `.guest` (the
  /// route returns guest as a first-class 200, never an error).
  func me() async throws -> MeSnapshot

  /// Ends the device session: best-effort server revoke, then always clears the
  /// Keychain token so the app returns to guest. Idempotent — calling it without
  /// a session is a no-op (M-AC-3.1).
  func signOut() async throws

  /// Permanently deletes the account and all its server data, then clears the
  /// local token (M-ACCT-US-6 / M-BR-ACCT-6). A transport/HTTP failure (other
  /// than an already-orphaned token) propagates so the UI does not falsely claim
  /// deletion.
  func deleteAccount() async throws
}

/// The result of a successful ``AuthService/verify(email:code:)``.
///
/// `created` distinguishes a first-time signup (`true`, M-AC-2.3) from a returning
/// login (`false`, M-AC-2.4); the UI may greet a new account differently.
struct Account: Equatable, Sendable {
  let email: String
  let created: Bool
}

/// The result of ``AuthService/me()`` — auth state plus the account's remembered
/// new-chat scope when signed in (mirrors web's `MeResult.lastUsedScope`).
struct MeSnapshot: Equatable, Sendable {
  let state: AuthState
  /// Present only when signed in and the account has a stored preference.
  let lastUsedScope: Format?

  static let guest = MeSnapshot(state: .guest, lastUsedScope: nil)

  static func signedIn(email: String, lastUsedScope: Format? = nil) -> MeSnapshot {
    MeSnapshot(state: .signedIn(email: email), lastUsedScope: lastUsedScope)
  }
}

/// Production ``AuthService`` over ``OakAPIClient`` (the network) and
/// ``TokenStore`` (the Keychain).
///
/// A value type holding two immutable actor references, so it is `Sendable`
/// without ceremony. It owns the **only** policy decisions the auth surface needs:
/// when to persist the token (on `verify`) and when to clear it (on `signOut` /
/// `deleteAccount`). Wire shapes are decoded into ``AuthVerifyResponse`` /
/// ``MeResponse``; error mapping happens inside ``OakAPIClient``.
struct LiveAuthService: AuthService {
  private let apiClient: OakAPIClient
  private let tokenStore: TokenStore

  init(apiClient: OakAPIClient, tokenStore: TokenStore) {
    self.apiClient = apiClient
    self.tokenStore = tokenStore
  }

  func requestCode(email: String) async throws {
    let endpoint = Endpoint(
      method: .post,
      path: "/api/auth/request-code",
      body: RequestCodeBody(email: email),
      requiresAuth: false
    )
    try await apiClient.sendNoContent(endpoint)
  }

  func verify(email: String, code: String) async throws -> Account {
    let endpoint = Endpoint(
      method: .post,
      path: "/api/auth/verify",
      body: VerifyBody(email: email, code: code),
      requiresAuth: false
    )
    let response = try await apiClient.send(endpoint, as: AuthVerifyResponse.self)
    // Persist the Bearer token (M-BR-ACCT-5) — the one place a token is stored.
    await tokenStore.set(response.token)
    return Account(email: response.email, created: response.created)
  }

  func me() async throws -> MeSnapshot {
    let endpoint = Endpoint(method: .get, path: "/api/auth/me", requiresAuth: true)
    let response = try await apiClient.send(endpoint, as: MeResponse.self)
    if response.signedIn, let email = response.email {
      let scope: Format? = response.lastUsedScope.map { Format(rawValue: $0) }.flatMap { format in
        // Drop `.unknown` — a retired wire value fails soft to nil so the chip
        // falls through to national-dex (server does the same via isFormat).
        if case .unknown = format { return nil }
        return format
      }
      return .signedIn(email: email, lastUsedScope: scope)
    }
    return .guest
  }

  func signOut() async throws {
    // Best-effort server revoke: a transport/HTTP failure must NOT block the
    // local return to guest, so it is logged (never swallowed silently) and the
    // token is cleared regardless. Idempotent (M-AC-3.1).
    let endpoint = Endpoint(method: .post, path: "/api/auth/signout", requiresAuth: true)
    do {
      try await apiClient.sendNoContent(endpoint)
    } catch {
      Log.auth.error("sign-out endpoint failed; clearing local token anyway")
    }
    await tokenStore.clear()
  }

  func deleteAccount() async throws {
    let endpoint = Endpoint(method: .delete, path: "/api/auth/account", requiresAuth: true)
    do {
      try await apiClient.sendNoContent(endpoint)
    } catch OakError.unauthorized {
      // The token is already orphaned (the account row is gone) — clearing the
      // local token completes the deletion from the device's point of view.
      Log.auth.info("account deletion saw 401 (already orphaned); clearing local token")
    }
    // Reached only on a confirmed 2xx delete or an already-orphaned 401: every
    // other failure (transport / 5xx / rate limit) propagated above with the
    // token intact, so the account is genuinely gone before we clear it.
    await tokenStore.clear()
  }
}

/// `POST /api/auth/request-code` body. `email` is identical on the wire, so no
/// `CodingKeys` are needed (the property name *is* the wire key).
private struct RequestCodeBody: Encodable, Sendable {
  let email: String
}

/// `POST /api/auth/verify` body. `email`/`code` are identical on the wire.
private struct VerifyBody: Encodable, Sendable {
  let email: String
  let code: String
}
